### PR TITLE
L4ping

### PR DESCRIPTION
**Pull request by MarSamMS**
Updated NPT to work with containers and added l4ping support

### CODE REVIEW

File: runPerftool.psm1 [hunk 1]
```diff
@@ -286,44 +286,342 @@ Function ProcessCommands{
...
+    [Switch] $RunOnContainers=$False,
+
+    [Parameter(Mandatory=$False, ParameterSetName="Default")]
+    [Parameter(Mandatory=$False, ParameterSetName="LocalTransmit")]
+    [Parameter(Mandatory=$False, ParameterSetName="RemoteTransmit")]
+    [Parameter(Mandatory=$True, ParameterSetName="RunOnContainers")]
+    [Parameter(Mandatory=$True, ParameterSetName="LocalTransmitOnContainers")]
+    [Parameter(Mandatory=$True, ParameterSetName="RemoteTransmitOnContainers")]
+    [Switch] $RunWithSinglePSSession=$False
+    )
```
Comment (amao9098) [inline runPerftool.psm1:481]: I'm a little confused here, especially about the LocalTransmit parameter. I think we should make as little change as possible to these script. I also think we should generalize the runOnContainers to runWithSSH since this is mainly adding SSH support if other people want to use it.

File: PERFTEST.PS1 [hunk 2]
```diff
@@ -8,11 +8,11 @@ Param(
     
     [ValidateSet("Default", "Azure", "Detail", "Max")]
     [parameter(Mandatory=$false)] [string] $Config = "Default",
-    [parameter(Mandatory=$false)]  [Array] $ToolList = @("ntttcp", "latte", "cps", "ctstraffic")
+    [parameter(Mandatory=$false)]  [Array] $ToolList = @("ntttcp", "latte", "l4ping", "cps", "ctstraffic")
 )
 
 $scriptName = $MyInvocation.MyCommand.Name 
-$SupportedTools = @("ntttcp", "latte", "cps", "ctstraffic")
+$SupportedTools = @("ntttcp", "latte", "l4ping", "cps", "ctstraffic")
```
Comment (DHarper89936) [inline PERFTEST.PS1:15]: nit: use single quotes

File: PERFTEST.PS1 [hunk 3]
```diff
@@ -8,11 +8,11 @@ Param(
     
     [ValidateSet("Default", "Azure", "Detail", "Max")]
     [parameter(Mandatory=$false)] [string] $Config = "Default",
-    [parameter(Mandatory=$false)]  [Array] $ToolList = @("ntttcp", "latte", "cps", "ctstraffic")
+    [parameter(Mandatory=$false)]  [Array] $ToolList = @("ntttcp", "latte", "l4ping", "cps", "ctstraffic")
```
Comment (DHarper89936) [inline PERFTEST.PS1:11]: nit: use single quotes

File: SetupTearDown.ps1 [hunk 4]
```diff
@@ -9,14 +9,26 @@
...
+
 .DESCRIPTION
     Run this script to setup your machine for PS Remoting so that you can leverage the functionality of runPerfTool.psm1
     Run this script at the end of the tool runs to restore state on the machines.
     Ex: SetupTearDown.ps1 -Setup or SetupTearDown.ps1 -Cleanup
 #>
 Param(
-    [switch] $Setup,
-    [switch] $Cleanup
+    [Parameter(Mandatory=$True, ParameterSetName="Setup")]              [switch] $Setup,
```
Comment (DHarper89936) [inline SetupTearDown.ps1:27]: These should not be mandatory (doing so would require you to specify -Cleanup:$false)

File: l4ping/l4ping.TESTGEN.ps1 [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    return $isValid
+} # validate_config()
+
+#===============================================
+# External Functions - Main Program
+#===============================================
+function test_main {
+    Param(
+        [parameter(Mandatory=$false)] [string] $Config = "Default",
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:108]: nit: single quote

File: l4ping/l4ping.TESTGEN.ps1 [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        [parameter(Mandatory=$true)]  [string] $DestIp,
+        [parameter(Mandatory=$true)]  [string] $SrcIp,
+        [parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = ""
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:111]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            Write-Output "l4ping$Config does not exist in .\l4ping\l4ping.Config.json. Please provide a valid config"
+            Throw
+        }
+        if (-Not (validate_config)) {
+            Write-Output "l4ping$Config is not a valid config"
+            Throw
+        }
+        [string] $g_DestIp     = $DestIp.Trim()
+        [string] $g_SrcIp      = $SrcIp.Trim()
+        [string] $dir          = (Join-Path -Path $OutDir -ChildPath "l4ping")
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:128]: nit: single quote

File: l4ping/l4ping.TESTGEN.ps1 [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        [string] $dir          = (Join-Path -Path $OutDir -ChildPath "l4ping") 
+        [string] $g_log        = "$dir\l4ping.Commands.txt"
+        [string] $g_logSend    = "$dir\l4ping.Commands.Send.txt"
+        [string] $g_logRecv    = "$dir\l4ping.Commands.Recv.txt"
+
+        New-Item -ItemType directory -Path $dir | Out-Null
+        
+        # Optional - Edit spaces in output path for Invoke-Expression compatibility
+        # $dir  = $dir  -replace ' ','` '
+        banner -Msg "l4ping Tests"
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:137]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        test_l4ping_generate -OutDir $dir
+    } catch {
+        Write-Output "Unable to generate l4ping commands"
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:140]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 10]
```diff
@@ -0,0 +1,143 @@
+#===============================================
+# Script Input Parameters Enforcement
+#===============================================
+Param(
+    [parameter(Mandatory=$false)] [string] $Config = "Default",
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:5]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    [parameter(Mandatory=$true)]  [string] $DestIp,
+    [parameter(Mandatory=$true)]  [string] $SrcIp,
+    [parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = ""
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:8]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    [parameter(Mandatory=$true)]  [string] $DestIp,
+    [parameter(Mandatory=$true)]  [string] $SrcIp,
+    [parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = "" 
+)
+$scriptName = $MyInvocation.MyCommand.Name 
+
+function input_display {
+    $g_path = Get-Location
+
+    Write-Output "============================================"
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:15]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    Write-Output "$g_path\$scriptName"
+    Write-Output " Inputs:"
+    Write-Output "  -Config     = $Config"
+    Write-Output "  -DestIp     = $DestIp"
+    Write-Output "  -SrcIp      = $SrcIp"
+    Write-Output "  -OutDir     = $OutDir"
+    Write-Output "============================================"
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:22]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+    Param(
+        [parameter(Mandatory=$true)] [String] $Msg
+    )
+    Write-Output "`n==========================================================================="
+    Write-Output "| $Msg"
+    Write-Output "==========================================================================="
+} # banner()
+
+function test_recv {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:39]: remove

File: l4ping/l4ping.TESTGEN.ps1 [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        [parameter(Mandatory=$true)]   [int] $ClientSendSize
+    )
+    [string] $cmd = "l4ping.exe -s $g_DestIp`:$Port -R $ClientSendSize -S $ClientReceiveSize"
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logRecv
+    Write-Output $cmd
+} # test_recv()
+
+function test_send {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:52]: Remove

File: l4ping/l4ping.TESTGEN.ps1 [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    [string] $out = (Join-Path -Path $OutDir -ChildPath $Fname)
+    [string] $cmd = "l4ping.exe -c $g_DestIp`:$Port -S $ClientSendSize -R $ClientReceiveSize -m $PingIterations -p `"$Percentiles`" -o $out"
+
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logSend
+    Write-Output $cmd
+} # test_send()
+
+function test_l4ping_generate {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:72]: Remove

File: l4ping/l4ping.TESTGEN.ps1 [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    Write-Output "  -SrcIp      = $SrcIp"
+    Write-Output "  -OutDir     = $OutDir"
+    Write-Output "============================================"
+} # input_display()
+
+#===============================================
+# Internal Functions
+#===============================================
+function banner {
+    [CmdletBinding()]
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:29]: Remove

File: SetupTearDown.ps1 [hunk 18]
```diff
@@ -29,6 +41,122 @@ Function SetupRemoting{
 
 } # SetupRemoting()
 
+Function SetupSshRemotingOnHost{
+
+    Write-Host "Enabling SSH Remoting on host computer..."
+
+    if (-NOT (Test-Path "$env:USERPROFILE\.ssh\id_ed25519.pub"))
```
[the remark this answers is not in the document]
Comment (DHarper89936) [reply SetupTearDown.ps1:48]: why not take the param for the .pub rather than hard code

File: SetupTearDown.ps1 [hunk 19]
```diff
@@ -29,6 +41,122 @@ Function SetupRemoting{
...
+    
+    Write-Host "`n`nRun the following command in each of the containers"
+    Write-Host ".\SetUpTearDown.ps1 -SetupContainer -AuthorizedKey '$authorizedKey'"
+
+} # SetupSshRemotingOnHost()
+
+
+Function SetupSshRemotingOnContainer{
+param(
+    [Parameter(Mandatory=$True)] [string]$AuthorizedKey
```
Comment (DHarper89936) [inline SetupTearDown.ps1:85]: Can this be an empty string?  If not, then add basic parameter validation
 [ValidateScript({-Not [String]::IsNullOrWhiteSpace($_)})]

File: PERFTEST.PS1 (hunk 3 above)
Comment (DHarper89936) [inline PERFTEST.PS1:11]: should there be a validation check for the list?

File: SetupTearDown.ps1 [hunk 20]
```diff
@@ -29,6 +41,122 @@ Function SetupRemoting{
 
 } # SetupRemoting()
 
+Function SetupSshRemotingOnHost{
```
Comment (DHarper89936) [inline SetupTearDown.ps1:44]: any reason we are hard-coding the .pub?  more extensible if we were to take in a parameter

File: SetupTearDown.ps1 [hunk 21]
```diff
@@ -49,21 +177,12 @@ Function CleanupRemoting{
 
 } # CleanupRemoting()
 
-#Main-function
```
Comment (DHarper89936) [inline SetupTearDown.ps1:52]: why the change in execution?  This change loses the catch of the exceptions.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 10 above)
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:5]: Parameter validation?

File: l4ping/l4ping.TESTGEN.ps1 [hunk 22]
```diff
@@ -0,0 +1,143 @@
+#===============================================
+# Script Input Parameters Enforcement
+#===============================================
+Param(
+    [parameter(Mandatory=$false)] [string] $Config = "Default",
+    [parameter(Mandatory=$true)]  [string] $DestIp,
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:6]: Parameter validation to make sure the string is a valid IP address.  Do we support both v4 and v6?

File: l4ping/l4ping.TESTGEN.ps1 [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+    [parameter(Mandatory=$true)]  [string] $SrcIp,
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:7]: Parameter validation to make sure the string is a valid IP address.

File: l4ping/l4ping.TESTGEN.ps1 [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+    [parameter(Mandatory=$true)]  [ValidateScript({Test-Path $_ -PathType Container})] [String] $OutDir = "" 
+)
+$scriptName = $MyInvocation.MyCommand.Name 
+
+function input_display {
+    $g_path = Get-Location
+
+    Write-Output "============================================"
+    Write-Output "$g_path\$scriptName"
+    Write-Output " Inputs:"
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:17]: nit: single quotes

File: l4ping/l4ping.TESTGEN.ps1 [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+    Write-Output "============================================"
+} # input_display()
+
+#===============================================
+# Internal Functions
+#===============================================
+function banner {
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)] [String] $Msg
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:31]: Parameter validation to not have allow an empty string

File: l4ping/l4ping.TESTGEN.ps1 [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+    )
+    Write-Output "`n==========================================================================="
+    Write-Output "| $Msg"
+    Write-Output "==========================================================================="
+} # banner()
+
+function test_recv {
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)]   [int] $Port,
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:41]: param validation ( 0-65535 )

File: l4ping/l4ping.TESTGEN.ps1 [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+        [parameter(Mandatory=$true)]   [int] $ClientReceiveSize,
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:42]: can we specify a valid range?

File: l4ping/l4ping.TESTGEN.ps1 [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+        [parameter(Mandatory=$true)]   [int] $ClientSendSize
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:43]: can we specify a valid range?

File: l4ping/l4ping.TESTGEN.ps1 [hunk 29]
```diff
@@ -0,0 +1,143 @@
...
+    [string] $cmd = "l4ping.exe -s $g_DestIp`:$Port -R $ClientSendSize -S $ClientReceiveSize"
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logRecv
+    Write-Output $cmd
+} # test_recv()
+
+function test_send {
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)]   [int]    $Port,
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:54]: validate (0-65535)

Comment (DHarper89936) [reply l4ping/l4ping.TESTGEN.ps1:54]: similar comments for rest of parameters below.  only allow what you know we support, validate any assumptions ( Output dir is present, etc)

File: l4ping/l4ping.TESTGEN.ps1 [hunk 30]
```diff
@@ -0,0 +1,143 @@
...
+
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_log
+    Write-Output $cmd | Out-File -Encoding ascii -Append $g_logSend
+    Write-Output $cmd
+} # test_send()
+
+function test_l4ping_generate {
+    [CmdletBinding()]
+    Param(
+        [parameter(Mandatory=$true)] [String] $OutDir
```
Comment (DHarper89936) [inline l4ping/l4ping.TESTGEN.ps1:74]: validate

File: l4ping/l4ping.TESTGEN.ps1 (hunk 5 above)
Comment (DHarper89936) [reply l4ping/l4ping.TESTGEN.ps1:108]: validate your parameters as much as possible

File: runPerftool.psm1
Comment (DHarper89936) [inline runPerftool.psm1:1]: Please make sure that these changes do not regress the non-container usage

**Review by DHarper89936 — CHANGES_REQUESTED** [review]
Please address comments

File: l4ping/l4ping.TESTGEN.ps1 (hunk 11 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:8]: Sure, I can do that. I copied this from the way all the other TestGen scripts are written. Would you like me to change all of them?

File: PERFTEST.PS1 (hunk 3 above)
Comment (MarSamMS) [reply PERFTEST.PS1:11]: I think that's what the normalize_toollist function does. Is that what you mean?

File: SetupTearDown.ps1 (hunk 19 above)
Comment (MarSamMS) [reply SetupTearDown.ps1:85]: I think making the parameter mandatory is already doing that. When I try to pass in empty string it gives me an error. Do you still want me to add the validation?

image

File: l4ping/l4ping.TESTGEN.ps1 (hunk 10 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:5]: Copied this from the way the other testgens are working. Would you like me to fix all of them?

File: l4ping/l4ping.TESTGEN.ps1 (hunk 22 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:6]: I think l4ping only supports v4 based on some code that I found [here](https://microsoft.visualstudio.com/OS/_git/os.2020?path=/onecore/vm/dv/net/tools/l4ping/PingApp.cpp&version=GBofficial/main&line=30&lineEnd=45&lineStartColumn=1&lineEndColumn=2&lineStyle=plain&_a=contents).

Should I fix all the TESTGEN scripts? They all seem to be missing this validation. If I just add it here, I think it could break the PerfTest.ps1 script.

File: l4ping/l4ping.TESTGEN.ps1 [hunk 31]
```diff
@@ -0,0 +1,143 @@
...
+        [parameter(Mandatory=$true)]   [int]    $ClientReceiveSize,
+        [parameter(Mandatory=$true)]   [int]    $ClientSendSize,
+        [parameter(Mandatory=$true)]   [int]    $PingIterations,
+        [parameter(Mandatory=$true)]   [String] $Percentiles,
+        [parameter(Mandatory=$true)]   [String] $OutDir,
+        [parameter(Mandatory=$true)]   [String] $Fname
+    )
+
+    [string] $out = (Join-Path -Path $OutDir -ChildPath $Fname)
+    [string] $cmd = "l4ping.exe -c $g_DestIp`:$Port -S $ClientSendSize -R $ClientReceiveSize -m $PingIterations -p `"$Percentiles`" -o $out"
```
[the remark this answers is not in the document]
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:64]: I don't think it's very useful. It just outputs how many times it restarts and then at the end it outputs the same percentile information that's in the output file. Do you think we should still capture it?

image

File: l4ping/l4ping.TESTGEN.ps1 [hunk 32]
```diff
@@ -0,0 +1,143 @@
...
+
+    for ($i=0; $i -lt $g_Config.Iterations; $i++) {
+        # vary port number
+        [int] $port = $g_Config.StartPort + ($i * $g_Config.Iterations)
+
+        # Output receive command
+        test_recv -Port $port -ClientReceiveSize $g_Config.ClientReceiveSize -ClientSendSize $g_Config.ClientSendSize
+
+        # Output send command
+        test_send -Port $port -ClientReceiveSize $g_Config.ClientReceiveSize -ClientSendSize $g_Config.ClientSendSize -PingIterations $g_Config.PingIterations -Percentiles $g_Config.Percentiles -OutDir $dir -Fname "l4ping$Config.iter$i.csv"
```
[the remark this answers is not in the document]
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:87]: Sure thing. Done.

File: SetupTearDown.ps1 (hunk 20 above)
Comment (MarSamMS) [reply SetupTearDown.ps1:44]: I was following the documentation [here](https://learn.microsoft.com/en-us/windows-server/administration/openssh/openssh_keymanagement#user-key-generation), but I can change it so the user passes in the key paths.

File: SetupTearDown.ps1 [hunk 33]
```diff
@@ -9,14 +9,26 @@
...
     Run this script to setup your machine for PS Remoting so that you can leverage the functionality of runPerfTool.psm1
     Run this script at the end of the tool runs to restore state on the machines.
     Ex: SetupTearDown.ps1 -Setup or SetupTearDown.ps1 -Cleanup
 #>
 Param(
-    [switch] $Setup,
-    [switch] $Cleanup
+    [Parameter(Mandatory=$True, ParameterSetName="Setup")]              [switch] $Setup,
+    [Parameter(Mandatory=$True, ParameterSetName="Cleanup")]            [switch] $Cleanup,
+    [Parameter(Mandatory=$True, ParameterSetName="SetupContainerHost")] [switch] $SetupContainerHost,
```
[the remark this answers is not in the document]
Comment (MarSamMS) [reply SetupTearDown.ps1:29]: Good catch. I meant to rename these parameters to match the functions. I can also fix them to not mention containers.

File: SetupTearDown.ps1 (hunk 18 above)
[the remark this answers is not in the document]
Comment (MarSamMS) [reply SetupTearDown.ps1:48]: Ok. I've removed the key generation. The user now needs to create the keys and pass in the paths/values to the script.

File: SetupTearDown.ps1 (hunk 21 above)
Comment (MarSamMS) [reply SetupTearDown.ps1:52]: Wasn't sure we needed the catch, since the script is ending after the catch anyway, but I've added it back.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 22 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:6]: Also, all the parameter validation is currently happening in validate_config to match how all the other TESTGEN scripts are doing it.

File: l4ping/l4ping.Config.json [hunk 34]
```diff
@@ -0,0 +1,11 @@
+{
```
[the remark this answers is not in the document]
Comment (MarSamMS) [reply l4ping/l4ping.Config.json:1]: Sure thing.

File: PERFTEST.PS1 (hunk 3 above)
Comment (MarSamMS) [reply PERFTEST.PS1:11]: Copied this from the way it was done in PerfTest before. Do you want me to fix all the old values as well?

File: l4ping/l4ping.TESTGEN.ps1 (hunk 10 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:5]: Line 9 in PerfTest.ps1 is validating the config value for all the TestGens. Would you like me to add validation here as well?

File: l4ping/l4ping.TESTGEN.ps1 (hunk 26 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:41]: Sure. I added this validation in the validate_config to match what all the other TestGens are doing.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 27 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:42]: I'm validating it's between 1 and 65,535 now

File: l4ping/l4ping.TESTGEN.ps1 (hunk 28 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:43]: I'm validating it's between 1 and 65,535 now

File: l4ping/l4ping.TESTGEN.ps1 (hunk 29 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:54]: Added validation in validate_config to match what the other TestGens are doing.

File: runPerftool.psm1 (hunk 1 above)
Comment (MarSamMS) [reply runPerftool.psm1:486]: Ok. I've updated the script to leave the original functions unchanged and added new functions that can handle both the old and new behaviors. So anyone using the old functions should be completely safe now.

I also replaced RunOnContainers with RunOverSSH and DisableFirewallConfiguration, which is the two things RunOnContainers was doing.

File: PERFTEST.PS1 (hunk 3 above)
Comment (DHarper89936) [reply PERFTEST.PS1]: It's a nit.  go ahead and leave as is.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 10 above)
Comment (DHarper89936) [reply l4ping/l4ping.TESTGEN.ps1:5]: It's a nit, go ahead and leave as is

File: SetupTearDown.ps1 (hunk 19 above)
Comment (DHarper89936) [reply SetupTearDown.ps1:81]: What about a string with spaces.  We should validate

Comment (MarSamMS) [reply SetupTearDown.ps1:81]: Ok. Fixed it.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 31 above)
[the remark this answers is not in the document]
Comment (amao9098) [reply l4ping/l4ping.TESTGEN.ps1:64]: no, I don't think it is necessary then. I was hoping for some output about each individual ping but it doesn't look like it captures it.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 23 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:7]: Added it.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 11 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:8]: Removed the default value here.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 17 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:29]: Leaving as is to match legacy behavior for all other TestGens

File: l4ping/l4ping.TESTGEN.ps1 (hunk 30 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:74]: Added it.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 5 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:147]: Copied validation from Perftest.ps1

File: l4ping/l4ping.TESTGEN.ps1 (hunk 6 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:150]: Done

**Comment by MarSamMS** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: PERFTEST.PS1 (hunk 2 above)
Comment (MarSamMS) [reply PERFTEST.PS1:15]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 7 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:167]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 8 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:176]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 9 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:179]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 12 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:15]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 13 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:22]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 14 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:39]: Leaving as is to match legacy behavior for all other TestGens

File: l4ping/l4ping.TESTGEN.ps1 (hunk 15 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:52]: Leaving as is to match legacy behavior for all other TestGens

File: l4ping/l4ping.TESTGEN.ps1 (hunk 16 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:72]: Leaving as is to match legacy behavior for all other TestGens

File: PERFTEST.PS1 (hunk 3 above)
Comment (MarSamMS) [reply PERFTEST.PS1:11]: Leaving as is based on offline discussion.

File: l4ping/l4ping.TESTGEN.ps1 (hunk 24 above)
Comment (MarSamMS) [reply l4ping/l4ping.TESTGEN.ps1:17]: Leaving as is based on offline discussion.

File: runPerftool.psm1
Comment (MarSamMS) [reply runPerftool.psm1:1]: Validated it's still working with both VMs and containers

**Review by DHarper89936 — APPROVED** [review]

File: runPerftool.psm1 [hunk 35]
```diff
@@ -325,6 +325,309 @@ Function ProcessCommands{
...
+    [Parameter(Mandatory=$True, ParameterSetName="RemoteTransmit", HelpMessage="Dest Machine Password?")]
+    [SecureString] $DestIpPassword,
+
+    [Parameter(Mandatory=$True, HelpMessage="Src Machine Username?")]
+    [string] $SrcIpUserName,
+
+    [Parameter(Mandatory=$True, ParameterSetName="Default", HelpMessage="Src Machine Password?")]
+    [Parameter(Mandatory=$True, ParameterSetName="LocalTransmit", HelpMessage="Src Machine Password?")]
+    [Parameter(Mandatory=$True, ParameterSetName="RemoteTransmit", HelpMessage="Src Machine Password?")]
+    [SecureString] $SrcIpPassword,
```
Comment (amao9098) [inline runPerftool.psm1:440]: is there a reason for not passing in the path to the ssh key?

**Review by amao9098 — APPROVED** [review]
